File: app/src/main/java/com/wellness/companion/domain/llm/ReflectionEngine.kt
package com.wellness.companion.domain.llm

import com.wellness.companion.domain.narrative.MirrorGenerator
import com.wellness.companion.data.db.entities.JournalEntry

class ReflectionEngine(private val engine: LlmEngine) {

    val isReady: Boolean get() = engine.isReady

    suspend fun reflect(title: String, body: String): List<String>? {
        if (!isReady) return null
        val prompt = "Title: $title\n\nContent: $body"
        val response = engine.generate(LlmPrompts.SOCRATIC_REFLECTION, prompt, 150, 0.9f)
        return response.split("\n")
            .map { it.trim().trimStart('1', '2', '3', '.', '-', ' ') }
            .filter { it.isNotBlank() && it.endsWith("?") }
    }

    suspend fun reframe(title: String, body: String): String? {
        if (!isReady) return null
        val prompt = "Title: $title\n\nContent: $body"
        return engine.generate(LlmPrompts.REFRAME_LENS, prompt, 100, 0.7f).trim()
    }

    suspend fun narrateMirror(mirror: MirrorGenerator.Mirror): String? {
        if (!isReady) return null
        val data = "Valence: ${mirror.avgValence}, Keywords: ${mirror.topWords.joinToString { it.first }}"
        return engine.generate(LlmPrompts.PATTERN_NARRATOR, data, 200, 0.5f).trim()
    }

    suspend fun contextualStarter(moodLabel: String?, timeOfDay: String): String? {
        if (!isReady) return null
        val prompt = "Mood: ${moodLabel ?: "Balanced"}, Time: $timeOfDay"
        return engine.generate(LlmPrompts.CONTEXTUAL_STARTER, prompt, 50, 1.0f).trim().trim('"')
    }

    suspend fun suggestTitle(body: String): String? {
        if (!isReady) return null
        return engine.generate(LlmPrompts.AUTO_TITLE, body, 20, 0.6f).trim().trim('"').trim('.')
    }

    suspend fun guidedQuestion(exchanges: List<Pair<String, String>>, context: String = ""): String? {
        if (!isReady) return null
        val user = buildString {
            if (context.isNotEmpty()) appendLine("Context: $context\n")
            if (exchanges.isEmpty()) {
                appendLine("Start the analytic session.")
            } else {
                appendLine("Recent exchanges:")
                exchanges.takeLast(3).forEach { (q, a) -> appendLine("Q: $q\nA: $a") }
                appendLine("\nAsk the next sharp question.")
            }
        }
        return engine.generate(LlmPrompts.GUIDED_QUESTION, user, 60, 0.8f).trim().trim('"')
    }

    suspend fun compileGuided(exchanges: List<Pair<String, String>>): String? {
        if (!isReady) return null
        val interview = exchanges.joinToString("\n") { "Q: ${it.first}\nA: ${it.second}" }
        return engine.generate(LlmPrompts.GUIDED_COMPILE, interview, 512, 0.7f).trim()
    }

    suspend fun synthesizeInsight(entries: List<JournalEntry>): String? {
        if (!isReady || entries.size < 2) return null
        val combined = entries.take(3).joinToString("\n---\n") { "${it.title}: ${it.body}" }
        val system = "Persona: The Analytic Observer. Task: Find the 'hidden thread' connecting these 3 notes. What is the user not saying directly? Constraint: Under 40 words. Be sharp."
        return engine.generate(system, combined, 100, 0.75f).trim()
    }
}

File: ios/WellnessCompanion/Domain/LLM/ReflectionEngine.swift
import Foundation

final class ReflectionEngine {
    private let engine: LlamaEngine

    init(engine: LlamaEngine) {
        self.engine = engine
    }

    var isReady: Bool { engine.isReady }

    func reflect(title: String, body: String) async -> [String]? {
        guard engine.isReady, !body.isEmpty else { return nil }
        // High temp for "sharp" lateral thinking
        let raw = await engine.generate(
            system: LlmPrompts.socraticReflection,
            user: "Entry: \(String(body.prefix(1000)))",
            maxTokens: 150, temperature: 0.9 
        )
        return raw.split(separator: "\n")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty && $0.hasSuffix("?") }
    }

    func guidedQuestion(exchanges: [(String, String)], context: String = "") async -> String? {
        guard engine.isReady else { return nil }
        var user = "Context: \(context)\n\n"
        if exchanges.isEmpty {
            user += "Ask the first sharp opening question."
        } else {
            // Only provide the last 3 exchanges to keep the AI focused and sharp
            let recent = exchanges.suffix(3)
            for (q, a) in recent { user += "Q: \(q)\nA: \(a)\n" }
            user += "\nAsk the next short, analytic question."
        }
        
        let raw = await engine.generate(
            system: LlmPrompts.guidedQuestion, user: user,
            maxTokens: 60, temperature: 0.8
        )
        return raw.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func compileGuided(exchanges: [(String, String)]) async -> String? {
        guard engine.isReady, !exchanges.isEmpty else { return nil }
        var user = "Dialogue for distillation:\n"
        for (q, a) in exchanges { user += "Q: \(q)\nA: \(a)\n" }
        
        let raw = await engine.generate(
            system: LlmPrompts.guidedCompile, user: user,
            maxTokens: 500, temperature: 0.7
        )
        return raw.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func narrateMirror(_ mirror: Mirror) async -> String? {
        guard engine.isReady else { return nil }
        let raw = await engine.generate(
            system: LlmPrompts.patternNarrator, 
            user: "Data: Valence \(mirror.avgValence), Keywords: \(mirror.topWords.map { $0.0 }.joined(separator: ", "))",
            maxTokens: 120, temperature: 0.5 // Lower temp for data consistency
        )
        return raw.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func contextualStarter(moodLabel: String?, timeOfDay: String) async -> String? {
        guard engine.isReady else { return nil }
        let raw = await engine.generate(
            system: LlmPrompts.contextualStarter, 
            user: "Mood: \(moodLabel ?? "Neutral"), Time: \(timeOfDay)",
            maxTokens: 50, temperature: 1.0 // Max creativity
        )
        return raw.trimmingCharacters(in: .whitespacesAndNewlines).trimmingCharacters(in: CharacterSet(charactersIn: "\""))
    }
    
    func suggestTitle(body: String) async -> String? {
        guard engine.isReady, !body.isEmpty else { return nil }
        let raw = await engine.generate(
            system: LlmPrompts.autoTitle,
            user: "Text: \(String(body.prefix(500)))",
            maxTokens: 20, temperature: 0.6
        )
        return raw.trimmingCharacters(in: .whitespacesAndNewlines).trimmingCharacters(in: CharacterSet(charactersIn: "\"."))
    }
}
